File: tests/ut/cpp/ops/test_ops_gelu.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include <memory>
#include "common/common_test.h"
#include "ir/dtype/type.h"
#include "ir/primitive.h"
#include "utils/tensor_construct_utils.h"
#include "abstract/dshape.h"
#include "abstract/abstract_value.h"
#include "ops/test_ops.h"
#include "ops/ops_func_impl/gelu.h"
#include "ops/test_value_utils.h"

namespace mindspore {
namespace ops {

struct GeLUShape {
  ShapeVector x_shape;
  ShapeVector out_shape;
};
struct GeLUType {
  TypePtr x_type;
  TypePtr out_type;
};

class TestGeLU : public TestOps, public testing::WithParamInterface<std::tuple<GeLUShape, GeLUType>> {};

TEST_P(TestGeLU, gelu_dyn_shape) {
  const auto &shape_param = std::get<0>(GetParam());
  const auto &dtype_param = std::get<1>(GetParam());

  GeLUFuncImpl gelu_func_impl;
  auto prim = std::make_shared<Primitive>("GeLU");
  auto x = std::make_shared<abstract::AbstractTensor>(dtype_param.x_type, shape_param.x_shape);
  auto expect_shape = std::make_shared<abstract::TensorShape>(shape_param.out_shape);
  auto expect_dtype = std::make_shared<TensorType>(dtype_param.out_type);

  auto out_shape = gelu_func_impl.InferShape(prim, {x});
  ASSERT_TRUE(*out_shape == *expect_shape);
  auto out_dtype = gelu_func_impl.InferType(prim, {x});
  ASSERT_TRUE(*out_dtype == *expect_dtype);
}

auto GeLUOpShapeTestCases = testing::ValuesIn({
  /* static */
  GeLUShape{{2, 3, 4}, {2, 3, 4}},
  /* dynamic shape */
  GeLUShape{{-1}, {-1}},
  GeLUShape{{-1, 2, 4}, {-1, 2, 4}},
  GeLUShape{{5, 3, -1, 2, 1}, {5, 3, -1, 2, 1}},
  GeLUShape{{5, 3, -1, 2, 1, 4, 7, 4}, {5, 3, -1, 2, 1, 4, 7, 4}},
  /* dynamic rank */
  GeLUShape{{-2}, {-2}},
});

auto GeLUOpTypeTestCases = testing::ValuesIn({
  GeLUType{kFloat16, kFloat16},
  GeLUType{kFloat32, kFloat32},
  GeLUType{kFloat64, kFloat64},
});

INSTANTIATE_TEST_CASE_P(TestGeLU, TestGeLU, testing::Combine(GeLUOpShapeTestCases, GeLUOpTypeTestCases));
}  // namespace ops
}  // namespace mindspore

File: tests/st/ops/test_ops_gelu.py
# Copyright 2024 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import numpy as np
import pytest

import mindspore as ms
from mindspore import ops
from mindspore.nn import Cell
from mindspore.ops import gelu

rtol = 1e-3


class GeluCell(Cell):
    def __init__(self):
        super().__init__()
        self.gelu = gelu

    def construct(self, x, approximate):
        return self.gelu(x, approximate)


@pytest.mark.level0
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_x86_ascend_training
@pytest.mark.parametrize('context_mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
@pytest.mark.parametrize('approximate', ['tanh', 'none'])
def test_ops_forward(context_mode, approximate):
    """
    Feature: test gelu forward
    Description: test gelu forward
    Expectation: success
    """
    ms.set_context(jit_level='O0')
    ms.context.set_context(mode=context_mode)

    gelu_cell = GeluCell()

    # 2 x 2
    x = np.array([[0.1, 0.2], [0.3, 0.4]], np.float32)

    output = gelu_cell(ms.tensor(x), approximate).asnumpy()
    if approximate:
        expect = np.array([[0.0540, 0.1159],
                           [0.1854, 0.2622]])
    else:
        expect = np.array([[0.0540, 0.1159],
                           [0.1854, 0.2622]])

    np.testing.assert_allclose(output, expect, rtol=rtol)

    gelu_cell.set_inputs(ms.tensor(shape=[None, None], dtype=ms.float32), approximate)

    # 3 x 3
    x = np.array([[0.1, 0.2], [0.3, 0.4]], np.float32)

    output = gelu_cell(ms.tensor(x), approximate).asnumpy()
    if approximate:
        expect = np.array([[0.0540, 0.1159],
                           [0.1854, 0.2622]])
    else:
        expect = np.array([[0.0540, 0.1159],
                           [0.1854, 0.2622]])

    np.testing.assert_allclose(output, expect, rtol=rtol)

    x = np.array([[0.1, 0.2, 0.3], [0.4, 0.5, 0.6], [0.7, 0.8, 0.9]], np.float32)

    output = gelu_cell(ms.tensor(x), approximate).asnumpy()
    if approximate:
        expect = np.array([[0.0540, 0.1159, 0.1854],
                           [0.2622, 0.3457, 0.4354],
                           [0.5306, 0.6304, 0.7342]])
    else:
        expect = np.array([[0.0540, 0.1159, 0.1854],
                           [0.2622, 0.3457, 0.4354],
                           [0.5306, 0.6305, 0.7343]])

    np.testing.assert_allclose(output, expect, rtol=rtol)

    gelu_cell.set_inputs(ms.tensor(shape=None, dtype=ms.float32), approximate)

    # 2 x 2 x 2
    x = np.array([[0.1, 0.2], [0.3, 0.4]], np.float32)

    output = gelu_cell(ms.tensor(x), approximate).asnumpy()
    if approximate:
        expect = np.array([[0.0540, 0.1159],
                           [0.1854, 0.2622]])
    else:
        expect = np.array([[0.0540, 0.1159],
                           [0.1854, 0.2622]])

    np.testing.assert_allclose(output, expect, rtol=rtol)

    x = np.array([[[0.1, 0.2], [0.3, 0.4]], [[0.5, 0.6], [0.7, 0.8]]], np.float32)

    output = gelu_cell(ms.tensor(x), approximate).asnumpy()
    if approximate:
        expect = np.array([[[0.0540, 0.1159],
                            [0.1854, 0.2622]],
                           [[0.3457, 0.4354],
                            [0.5306, 0.6304]]])
    else:
        expect = np.array([[[0.0540, 0.1159],
                            [0.1854, 0.2622]],
                           [[0.3457, 0.4354],
                            [0.5306, 0.6305]]])

    np.testing.assert_allclose(output, expect, rtol=rtol)


@pytest.mark.level1
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_arm_ascend_training
@pytest.mark.parametrize('context_mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
@pytest.mark.parametrize('approximate', ['tanh', 'none'])
def test_ops_backward(context_mode, approximate):
    """
    Feature: test gelu backward
    Description: test gelu backward
    Expectation: success
    """
    ms.set_context(jit_level='O0')
    ms.context.set_context(mode=context_mode)

    gelu_cell = GeluCell()

    # 2 x 2
    x = np.array([[0.1, 0.2], [0.3, 0.4]], np.float32)

    output = ops.grad(gelu_cell, (0))(ms.tensor(x), approximate).asnumpy()
    if approximate:
        expect = np.array([[0.5795, 0.6575],
                           [0.7323, 0.8027]])
    else:
        expect = np.array([[0.5795, 0.6575],
                           [0.7323, 0.8027]])

    np.testing.assert_allclose(output, expect, rtol=rtol)
